### ISSUE TITLE
Uso más eficiente del mapa de colisiones

### ISSUE THREAD

**Issue by GoogleCodeExporter** [ludica-squamata/mano-gift#44]
```
Estaba pensando que nosotros usamos dos colores (bien distintos) para almacenar 
dos datos en el mapa de colisiones (si es o no transitable). Si bien es todo lo 
que necesitamos por ahora, se me ocurre que más adelanete podríamos necesitar 
indicar que una zona es, por ejemplo, transitable pero dificultosa, o con 
elevación, o parcialmente sumergida en agua. Se me ocurrió entonces que 
podríamos usar un solo color para definir eso, y usar los otros dos colores 
para definir otras cosas. Todo ello pensando en que podemos determinar los 
valores RGB exactos de un punto y creo también de un cuadro, sin demasiada 
dificultad (más aun si hacemos los mapas teniendo en cuenta la grilla, de modo 
que todo un cuadro tenga el mismo color).

Podríamos decir, por ejempo, que un valor Azul de 0 es intransitable, y 1 es 
simplemente transitable, y otras variables podrían tener otros valores. Lo 
mismo para el rojo o el verde (y no sé si no podemos calular tambien el valor 
de alpha: ahi serían 4 valores).

Por supuesto, la imagen resultante sería bastante dificil de leer, pero bleh! 
Tiene que ser coherente para la computadora, no para un ser humano (asumiendo 
claro, cosa que tampoco creo que sea muy dificil de hacer, que la imagen ésta 
sea creada por el generador: me consta que puede guardar en png [pygame puede]).
```

Original issue reported on code.google.com by `daniel.e.rossy` on 9 Dec 2013 at 12:17

**Comment by zenieldanaku** [comment]
Se me acaba de ocurrir que si, al escanear el mapa de colisiones (los 1024 px), pasamos punto por punto, también podemos indicar la posición de los props mediante un punto. En sí ya hacemos algo de esto, considerando que las colisiones son en realidad los propios props, pero no sé, se me ocurre ahora.

**Comment by zenieldanaku** [comment]
Hoy agregué la capacidad auditiva a los mobs (#35) y se me ocurría que podemos agregarle al mapa de colisiones un nuevo color, que indique que una zona es mas "ruidosa" que el suelo normal (algo así como caminar sobre hojas quebradizas).

**Comment by einacio** [comment]
O también atenuada, como una alfombra. Yo diría que sea un multiplicador,
que podríamos encodearlo incluso en el mismo color.

Ejemplo, supongamos que el canal A es el factor, los primeros 3 son el
indicador de zona y el cuarto es un multiplicador, con A=100 normal y podes
atenuar hasta 0 o amplificar hasta x2.55.

También podemos darle a cada canal el significado, en vez de trabajar por
colores. Nos puede dar mucha variedad, pero es una lógica más compleja

**Comment by zenieldanaku** [comment]
No entiendo a qué te referís con "canal".

**Comment by einacio** [comment]
Un color tiene 3 o 4 canales o componentes, R G B y A si esta habilitado.
Se pueden leer por separado para cada píxel, y así tener lógica más
compleja, más si usamos flags binarias.
La idea la leí en el juego Monaco, que usa imágenes para savegames
compartibles o algo asi

**Comment by zenieldanaku** [comment]
Si, calculé que venia por ese lado la mano (fue lo que había dicho yo, varios comentarios atrás).

O sea, vos decis usar R como x, G como y, y B como z?

**Comment by einacio** [comment]
No, yo decía usarlos para representar diferentes cosas, por ejemplo en R
marcas si un terreno es transitable, en B el multiplicador de ruido, etc.
Aunque si tenemos pocas cosas que marcar, usar un color entero es más
fácil, y para lo del ruido simplemente lo hacemos con un color +alpha o 4/5
colores para niveles prefijados

**Comment by zenieldanaku** [comment]
_Por ahora_ tenemos pocas cosas que marcar, sí, pero me parece que estaría bueno dejar la puerta abierta para ampliar el sistema más adelante de alguna manera.

**Comment by zenieldanaku** [comment]
Hoy me di cuenta de que la posición es un dato implícito del pixel, y que el color de un pixel sería el resultante de todos los metadatos de ese pixel (toda la información encodeada), por lo que habría que fijarse bien que las combinaciones, al menos para la computadora, fuesen únicas (te envié un mail al respecto)

**Comment by zenieldanaku** [comment]
Se me acaba de ocurrir, mirando el mapa viejo de casas y árboles, que en vez de colocar la posición de cada Sprite individualmente en el json, se podría asignar un color al area en el mapa de colisión donde aparecen esos props, y luego indicar en el json que color corresponde a que prop. No podemos hacer lo mismo con las Salidas porque, por ejemplo, en un área de agua (inicialmente inaccesible) podría haber una salida (aunque en ese ejemplo en particular el area de colisión podría estar rodeando la salida, but I digress)

**Comment by einacio** [comment]
eso es una cuestion de conveniencia realmente. el json es mas facil de editar a mano, lo podes hacer desde el mismo IDE donde estes trabajando. editar una imagen requiere utilizar otro editor.

---
Respecto de lo otro, nada evita que en agua haya una salida. a menos que uses imagenes de paleta, normalmente estarias usando RGB, lo cual te permite tener 3 capas de definiciones, podes marcar por ejemplo escenario en la capa R, sprites en la capa G y salidas/entradas en la capa B
incluso podes usar RGBA si quisieras una cuarta capa o 255 definiciones por capa no te alcanzan

**Comment by zenieldanaku** [comment]
Si, si. Lo de usar las 3 capas ya lo habíamos hablado en este mismo issue más arriba. Por eso terminé esa sección diciendo "but I digress": me estaba yendo de tema.

**Comment by danielrossyamitrano** [comment]
> Un efecto de esto sería que la mascara resultante ocuparía todo el tamaño del mapa, con todos sus bits, salvo los que representan a la salida, seteados en 0. No sé si eso no sería deperdiciar información.
Bueno no. Creo que con un poco de esfuerzo (PixelArray) se pueden extraer los canales individuales de los que luego se extraería la capa de bits.

**Comment by danielrossyamitrano** [comment]
> >     Un efecto de esto sería que la mascara resultante ocuparía todo el tamaño del mapa, con todos sus bits, 
> >     salvo los que   representan a la salida, seteados en 0. No sé si eso no sería deperdiciar información.

      

>    Bueno no. Creo que con un poco de esfuerzo (PixelArray) se pueden extraer los canales individuales de los 
>    que luego se extraería la capa de bits. Incluso se me ocurre que, luego de extraídos, se pueden re-combinar
>    de formas diferentes (Rojo con verde, o Azul y rojo, cosas así, o Rojo + Alpha incluso) para luego extraer
>    máscaras de bits de ESAS combinaciones... pero no sé qué tanto proceso sería todo eso porque trabajar con 
>    PixeArray es laborioso para mi computadora por lo menos.

[Tal vez no tanto...](https://github.com/danielrossyamitrano/_mis_pruebas/tree/master/extract)

**Comment by danielrossyamitrano** [comment]
Una forma de hacer uso de las 8 sombras sin crear PixelArrays para las montañas, es crear las sombras en el mismo dibujo de la montaña (como era originaaaaaaalmente) y definir un "área de colisión" en la que los mobs, si se paran encima se _oscurecen_, como si estuvieran _tapados por la sombra_.
